qa/tasks/cephfs/test_strays.py

```diff
@@ -1022,3 +1022,35 @@ def test_replicated_delete_speed(self):
 
         duration = (end - begin).total_seconds()
         self.assertLess(duration, (file_count * tick_period) * 0.25)
+    
+    def test_asok_dump_stray_command(self):
+        """
+        Test MDS asok dump stray command
+        """
+
+        LOW_LIMIT = 50
+        # need to create more folder to force fragmentation, creating more then needed
+        # to be on the safe side.
+        # we want to test the case when dumping stray folder must wait for the next dirfrag to be fetched
+        NUM_DIRS = LOW_LIMIT * 20
+        TOP_DIR = "topdir"
+        self.config_set("mds", "mds_bal_split_size", str(LOW_LIMIT))
+        self.assertEqual(self.config_get("mds", "mds_bal_split_size"), str(LOW_LIMIT), "LOW_LIMIT was not set on mds!")
+
+         # enable snaps
+        self.fs.set_allow_new_snaps(True)
+      
+        # create 2 level tree with enough folders to force the stray folder be fragmented
+        # total of NUM_DIRS subdirs will be created
+        self.mount_a.run_shell(f"mkdir -p {TOP_DIR}/subdir{{1..{NUM_DIRS}}}")  
+        # create snapshot
+        self.mount_a.run_shell(f"mkdir {TOP_DIR}/.snap/snap1")
+
+        # delete 2nd level dirs to generate strays
+        # don't wait, we want to dump stray dir while delete runs, to make it more interesting
+        self.mount_a.run_shell(f"rm -rf {TOP_DIR}/*", wait=False)
+
+        # wait for all deleted folders to become strays
+        self.wait_until_equal(
+        lambda: len(self.fs.rank_tell(["dump", "stray"])),
+        expect_val=NUM_DIRS, timeout=60, period=1)
```

Comment:
Testing this with 1000 strays is a good functional test, but can we also test this with a much larger number of strays? E.g. 10^6. (BTW someone on ceph-users said today they have 10^8 strays currently).

I'm concerned about us adding new mds `dump *` commands which have unbounded output. Operators usually use these commands when things are not going well, so it's pretty likely someone will run `dump strays` when they have a huge number of strays. We need to make sure `dump strays` is not harmful to the MDS in this case.

Ideas that would be acceptable imho include: paginate, refuse if num_strays>~1000, timeout the function internally, ...

Reply:
Thanks for this note. The main problem with any kind of solution mentioned - we don't know the number of strays in advance. We need to scan the tree and to find it out. Scanning the tree requires taking MDS log, and it's the same way we do when printing strays.
Maybe we can think about having a counter for approximate num of strays in future.
You can open a tracker item for this issue

Reply:
Let's have the functionality in place (we can defer the backport till we find an effective way to not worsen the situation when dumping huge stray counts).